anndata/tests/test_concatenate_disk.py

```diff
@@ -0,0 +1,853 @@
+def _filled_array_np(a, fill_value=None):
+    if fill_value is None:
+        fill_value = np.nan
+    return np.broadcast_to(fill_value, a.shape)
+
+
+@filled_like.register(DaskArray)
+def _filled_array(a, fill_value=None):
+    return as_dense_dask_array(_filled_array_np(a, fill_value))
+
+
+@filled_like.register(sparse.spmatrix)
+def _filled_sparse(a, fill_value=None):
+    if fill_value is None:
+        return sparse.csr_matrix(a.shape)
+    else:
+        return sparse.csr_matrix(np.broadcast_to(fill_value, a.shape))
+
+
+@filled_like.register(pd.DataFrame)
+def _filled_df(a, fill_value=np.nan):
+    # dtype from pd.concat can be unintuitive, this returns something close enough
+    return a.loc[[], :].reindex(index=a.index, fill_value=fill_value)
+
+
+def check_filled_like(x, fill_value=None, elem_name=None):
+    if fill_value is None:
+        assert_equal(x, filled_like(x), elem_name=elem_name)
+    else:
+        assert_equal(x, filled_like(x, fill_value=fill_value), elem_name=elem_name)
+
+
+def make_idx_tuple(idx, axis):
+    tup = [slice(None), slice(None)]
+    tup[axis] = idx
+    return tuple(tup)
+
+
+# Will call func(sparse_matrix) so these types should be sparse compatible
+# See array_type if only dense arrays are expected as input.
+@pytest.fixture(
+    params=[asarray, sparse.csr_matrix, sparse.csc_matrix, as_dense_dask_array],
+    ids=["np_array", "scipy_csr", "scipy_csc", "dask_array"],
+)
+def array_type(request):
+    return request.param
+
+
+@pytest.fixture(params=["inner", "outer"])
+def join_type(request):
+    return request.param
+
+
+@pytest.fixture(params=[0, np.nan, np.pi])
+def fill_val(request):
+    return request.param
+
+
+@pytest.fixture(params=[0, 1])
+def axis(request):
+    return request.param
+
+
+@pytest.fixture(params=list(merge.MERGE_STRATEGIES.keys()))
+def merge_strategy(request):
+    return request.param
+
+
+def fix_known_differences(orig, result, backwards_compat=True):
+    """
+    Helper function for reducing anndata's to only the elements we expect to be
+    equivalent after concatenation.
+
+    Only for the case where orig is the ground truth result of what concatenation should be.
+
+    If backwards_compat, checks against what `AnnData.concatenate` could do. Otherwise checks for `concat`.
+    """
+    orig = orig.copy()
+    result = result.copy()
+
+    result.strings_to_categoricals()  # Should this be implicit in concatenation?
+
+    # TODO
+    # * merge varm, varp similar to uns
+    # * merge obsp, but some information should be lost
+    del orig.obsp  # TODO
+
+    if backwards_compat:
+        del orig.varm
+        del orig.varp
+        result.obs.drop(columns=["batch"], inplace=True)
+
+    # Possibly need to fix this, ordered categoricals lose orderedness
+    for k, dtype in orig.obs.dtypes.items():
+        if is_categorical_dtype(dtype) and dtype.ordered:
+            result.obs[k] = result.obs[k].astype(dtype)
+
+    return orig, result
```

Comment:
I don't think this exact function will be relevant here, since I don't think we'll implement the `AnnData.concatenate` API